knora/excel2xml.py

```diff
@@ -0,0 +1,2127 @@
+) -> etree._Element:
+    """
+    Make a <color-prop> from one or more colors. The color(s) can be provided as string or as PropertyElement with a
+    string inside. If provided as string, the permission for every value is "prop-default".
+
+    To create one ``<color>`` child, use the param ``value``, to create more than one ``<color>`` children, use
+    ``values``.
+
+    Args:
+        name: the name of this property as defined in the onto
+        value: a string/PropertyElement, or an iterable of identical strings/PropertyElements
+        values: an iterable of (usually distinct) strings/PropertyElements
+        calling_resource: the name of the parent resource (for better error messages)
+
+    Examples:
+        >>> make_color_prop(":testproperty", "#00ff66")
+                <color-prop name=":testproperty">
+                    <color permissions="prop-default">#00ff66</color>
+                </color-prop>
+        >>> make_color_prop(":testproperty", PropertyElement("#00ff66", permissions="prop-restricted", comment="example"))
+                <color-prop name=":testproperty">
+                    <color permissions="prop-restricted" comment="example">#00ff66</color>
+                </color-prop>
+        >>> make_color_prop(":testproperty", values=["#00ff66", "#000000"])
+                <color-prop name=":testproperty">
+                    <color permissions="prop-default">#00ff66</color>
+                    <color permissions="prop-default">#000000</color>
+                </color-prop>
+        >>> make_color_prop(":testproperty", value=["#00ff66", "#000000"])
+                ERROR. Use this if some fields in your data source should be equal, but you cannot trust
+                your source. This method call will only work if all items of "value" are identical.
+
+    See https://docs.dasch.swiss/latest/DSP-TOOLS/dsp-tools-xmlupload/#color-prop
+    """
+
+    # check the input: prepare a list with valid values
+    values_new = _check_and_prepare_values(
+        value=value,
+        values=values,
+        name=name,
+        calling_resource=calling_resource
+    )
+
+    # check value type
+    for val in values_new:
+        if not re.search(r"^#[0-9a-f]{6}$", str(val.value).strip(), flags=re.IGNORECASE):
+            raise BaseError(f"Invalid color format for prop '{name}' in resource '{calling_resource}': '{val.value}'")
+
+    # make xml structure of the value
+    prop_ = etree.Element(
+        "{%s}color-prop" % (xml_namespace_map[None]),
+        name=name,
+        nsmap=xml_namespace_map
+    )
+    for val in values_new:
+        kwargs = {"permissions": val.permissions}
+        if check_notna(val.comment):
+            kwargs["comment"] = val.comment
+        value_ = etree.Element(
+            "{%s}color" % (xml_namespace_map[None]),
+            **kwargs,
+            nsmap=xml_namespace_map
+        )
+        value_.text = str(val.value).strip()
+        prop_.append(value_)
+
+    return prop_
```

Comment:
there is a lot of code repetition between those `make_XY_prop()` functions.

They always seem to do:
- perform `_check_and_prepare_values()`
- loop over the new values, apply a regex and throw an exception if the regex doesn't match
- create the prop_ et.Element
- loop over the values again, add the value subelements to the prop_
- return the prop_

I wonder if that repetition could be extracted elegantly